src/java.base/share/classes/jdk/internal/loader/SynchronizedLoader.java

```diff
@@ -0,0 +1,152 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.loader;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.lang.ClassLoader;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+/**
+ * Defines a version of loadClass that handles synchronization on the ClassLoader object
+ * for applications that release the ClassLoader lock but still expect the multiple
+ * threads to allow the first thread to successfully load the class and the other threads
+ * to wait then also succeed.
+ * This is for compatibility with long standing VM behavior, and may be deprecated in future releases.
+ **/
+public class SynchronizedLoader {
+
+    private static class NameLoaderRef {
+        private String name;
+        private ClassLoader loader;
+        public NameLoaderRef(String name, ClassLoader loader) {
+            this.name = name;
+            this.loader = loader;
+        }
+        @Override
+        public int hashCode() { return loader.hashCode() ^ name.hashCode(); }
+
+        public boolean equals(NameLoaderRef key) {
+            return key.name.equals(this.name) && key.loader == this.loader;
+        }
+        @Override
+        public boolean equals(Object obj) {
+            assert (obj instanceof NameLoaderRef);
+            return equals((NameLoaderRef) obj);
+        }
+
+        public ClassLoader loader() { return loader; }
+    }
+
+    // Synchronization for class loading when the current class loader is NOT
+    // parallel capable.  The non-parallel capable class loader locks
+    // the ClassLoader object. This maps a class name to the thread that is currently
+    // loading the class in case the ClassLoader lock is broken.
+    private static final ConcurrentHashMap<NameLoaderRef, Thread> threadLoadingClassMap
+        = new ConcurrentHashMap<>();
+
+    // The threadLoadingClassMap is a concurrent hashtable that keeps track of the threads
+    // that are currently loading each class by this class loader.  The class name is mapped to
+    // the thread that is first seen loading that class, then removed when loading is complete
+    // for that class.
+    private static final Thread threadLoadingClass(NameLoaderRef ref, Thread thread) {
+        return threadLoadingClassMap.putIfAbsent(ref, thread);
+    }
+
+    private static final void removeThreadLoadingClass(NameLoaderRef ref) {
+        threadLoadingClassMap.remove(ref);
+        // Notify threads waiting on the class loader lock
+        // that this class has been loaded or failed.
+        ref.loader().notifyAll();
+    }
```

Comment:
This can use `jdk.internal.loader.ClassLoaderValue` instead:
```java
// Synchronization for class loading when the current class loader is NOT
// parallel capable.  The non-parallel capable class loader locks
// the ClassLoader object. This maps a class name to the thread that is currently
// loading the class in case the ClassLoader lock is broken.
private static final ClassLoaderValue<Thread> threadLoadingClassMap = new ClassLoaderValue<>();

// The threadLoadingClassMap is a concurrent hashtable that keeps track of the threads
// that are currently loading each class by this class loader.  The class name is mapped to
// the thread that is first seen loading that class, then removed when loading is complete
// for that class.
private static final Thread threadLoadingClass(ClassLoader loader, String name, Thread thread) {
	return threadLoadingClassMap.sub(name).putIfAbsent(loader, thread);
}

private static final void removeThreadLoadingClass(ClassLoader loader, String name) {
	threadLoadingClassMap.sub(name).remove(loader);
	// Notify threads waiting on the class loader lock
	// that this class has been loaded or failed.
	loader.notifyAll();
}
```

Reply:
Hi Exe-Boss - are you a Java user and do you work on code that does "interesting" class loading?  Thanks for looking at this - I really appreciate it.  I don't know ClassLoaderValue at all (reading the code last night didn't help me) but I'll look into your suggestion.  Thank you.

Reply:
Do you work on JBOSS?

Reply:
> do you work on code that does "interesting" class loading?

The closest thing would have been the **Minecraft** wrappers for **[NOVA‑Core]**.

[NOVA‑Core]: https://github.com/NOVA-Team/NOVA-Monorepo

--------------------------------------------------------------------------------

> Do you work on JBOSS?

I have no idea what that is.